File: packages/android-kotlin/src/main/kotlin/com/onchainux/walletconnect/WCClient.kt
/**
 * WCClient — WalletConnect v2 client for Android using WalletConnectKotlin SDK.
 *
 * Provides native Kotlin implementation of the WalletConnect v2 protocol:
 * - Pairing URI generation and QR code display
 * - Session proposal and establishment via WalletConnectKotlin SDK
 * - X25519 key exchange (handled by SDK)
 * - Relay connection (Waku, handled by SDK)
 * - JSON-RPC method dispatch (eth_sendTransaction, personal_sign, etc.)
 * - Balance fetching via eth_getBalance
 * - SIWE signing via personal_sign
 *
 * Uses the official WalletConnectKotlin SDK (com.walletconnect:sign).
 */
package com.cinacoin.walletconnect

import android.content.Context
import com.cinacoin.core.AppMetadata
import com.walletconnect.android.Core
import com.walletconnect.android.CoreClient
import com.walletconnect.android.internal.common.model.Server
import com.walletconnect.sign.Sign
import com.walletconnect.sign.SignClient
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.*
import java.util.UUID
import org.json.JSONArray
import org.json.JSONObject

// ============================================================
// Status
// ============================================================

sealed class WCStatus {
    object Disconnected : WCStatus()
    object Connecting : WCStatus()
    object Pairing : WCStatus()
    object Connected : WCStatus()
    data class Error(val message: String) : WCStatus()
}

// ============================================================
// Session
// ============================================================

data class WCSession(
    val topic: String,
    val peerMetadata: Map<String, String> = emptyMap(),
    val accounts: List<String> = emptyList(),
    val namespaces: Map<String, Any> = emptyMap(),
    val relayProtocol: String = "waku"
)

// ============================================================
// Transaction Request
// ============================================================

data class WCTransactionRequest(
    val from: String,
    val to: String,
    val value: String? = null,
    val data: String? = null,
    val gas: String? = null,
    val gasPrice: String? = null,
    val maxFeePerGas: String? = null,
    val maxPriorityFeePerGas: String? = null,
    val nonce: String? = null,
    val chainId: Int? = null
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("from", from)
        put("to", to)
        value?.let { put("value", it) }
        data?.let { put("data", it) }
        gas?.let { put("gas", it) }
        gasPrice?.let { put("gasPrice", it) }
        maxFeePerGas?.let { put("maxFeePerGas", it) }
        maxPriorityFeePerGas?.let { put("maxPriorityFeePerGas", it) }
        nonce?.let { put("nonce", it) }
        chainId?.let { put("chainId", chainId) }
    }
}

// ============================================================
// Events
// ============================================================

sealed class WCEvent {
    object Disconnected : WCEvent()
    data class Connected(val session: WCSession) : WCEvent()
    data class SessionUpdate(val session: WCSession?) : WCEvent()
    data class Error(val error: Throwable) : WCEvent()
}

// ============================================================
// Methods & Events Constants
// ============================================================

object WCMethods {
    const val ETH_SEND_TRANSACTION = "eth_sendTransaction"
    const val ETH_SIGN_TRANSACTION = "eth_signTransaction"
    const val PERSONAL_SIGN = "personal_sign"
    const val ETH_SIGN_TYPED_DATA = "eth_signTypedData"
    const val ETH_SIGN_TYPED_DATA_V4 = "eth_signTypedData_v4"
    const val WALLET_SWITCH_ETHEREUM_CHAIN = "wallet_switchEthereumChain"
    const val WALLET_ADD_ETHEREUM_CHAIN = "wallet_addEthereumChain"
    const val ETH_ACCOUNTS = "eth_accounts"
    const val ETH_CHAIN_ID = "eth_chainId"
    const val ETH_GET_BALANCE = "eth_getBalance"

    val standardEvmMethods = listOf(
        ETH_SEND_TRANSACTION, ETH_SIGN_TRANSACTION, PERSONAL_SIGN,
        ETH_SIGN_TYPED_DATA, ETH_SIGN_TYPED_DATA_V4,
        WALLET_SWITCH_ETHEREUM_CHAIN, WALLET_ADD_ETHEREUM_CHAIN,
        ETH_ACCOUNTS, ETH_CHAIN_ID
    )
}

object WCEvents {
    const val CHAIN_CHANGED = "chainChanged"
    const val ACCOUNTS_CHANGED = "accountsChanged"
    val standardEvmEvents = listOf(CHAIN_CHANGED, ACCOUNTS_CHANGED)
}

// ============================================================
// Errors
// ============================================================

sealed class WCError(message: String) : Exception(message) {
    object NotConnected : WCError("Not connected to a wallet")
    object NotConfigured : WCError("WCClient has not been configured")
    data class InvalidUrl(val url: String) : WCError("Invalid relay URL: $url")
    object SessionTimeout : WCError("Session establishment timed out")
    data class RequestTimeout(val method: String) : WCError("Request '$method' timed out")
    data class RpcError(val code: Int, val message: String) : WCError("RPC error $code: $message")
    object InvalidProposalResponse : WCError("Invalid session proposal response")
    data class RelayError(val message: String) : WCError("Relay error: $message")
    data class CryptoError(val message: String) : WCError("Crypto error: $message")
}

// ============================================================
// Parsed WC URI
// ============================================================

data class ParsedWcUri(
    val topic: String,
    val relayProtocol: String,
    val relayUrl: String,
    val symKey: String
)

// ============================================================
// WC Client — real WalletConnectKotlin SDK integration
// ============================================================

/**
 * WalletConnect v2 client for Android native integration.
 *
 * Wraps the official WalletConnectKotlin SDK (com.walletconnect:sign)
 * for real session management, pairing, and JSON-RPC communication.
 */
class WCClient {

    companion object {
        @Volatile
        private var INSTANCE: WCClient? = null

        fun getInstance(): WCClient =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: WCClient().also { INSTANCE = it }
            }
    }

    // State flows
    private val _status = MutableStateFlow<WCStatus>(WCStatus.Disconnected)
    val status: StateFlow<WCStatus> = _status.asStateFlow()

    private val _session = MutableStateFlow<WCSession?>(null)
    val session: StateFlow<WCSession?> = _session.asStateFlow()

    private val _pairingUri = MutableStateFlow<String?>(null)
    val pairingUri: StateFlow<String?> = _pairingUri.asStateFlow()

    private val _accounts = MutableStateFlow<List<String>>(emptyList())
    val accounts: StateFlow<List<String>> = _accounts.asStateFlow()

    private val _events = MutableSharedFlow<WCEvent>(extraBufferCapacity = 64)
    val events: Flow<WCEvent> = _events.asSharedFlow()

    // Configuration
    var relayUrl: String = "wss://relay.walletconnect.com"
    var projectId: String = ""
    var metadata: AppMetadata? = null
    var requiredChains: List<String> = listOf("eip155:1")
    var requiredMethods: List<String> = WCMethods.standardEvmMethods
    var requiredEvents: List<String> = WCEvents.standardEvmEvents

    // Internal state
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var isInitialized = false

    // WalletConnectKotlin SDK clients
    private var coreClient: CoreClient? = null
    private var signClient: SignClient? = null

    // Pending request tracking
    private var nextRequestId: Long = 1
    private val pendingRequests = mutableMapOf<Long, CompletableDeferred<JSONObject>>()

    // Event listener IDs from SDK
    private var sessionProposalListenerId: Long? = null
    private var sessionRequestListenerId: Long? = null
    private var sessionDeleteListenerId: Long? = null
    private var sessionSettleListenerId: Long? = null

    /**
     * Initialize the WalletConnectKotlin SDK.
     * Call this once at app startup before any WC operations.
     */
    fun initialize(
        context: Context,
        projectId: String,
        metadata: AppMetadata,
        server: Server = Server("relay.walletconnect.com")
    ) {
        if (isInitialized) return

        this.projectId = projectId
        this.metadata = metadata

        // Initialize the Core SDK
        val core = CoreClient(
            context = context,
            projectId = projectId,
            server = server
        )
        coreClient = core
        core.initialize()

        // Initialize the Sign SDK
        val sign = SignClient(
            core = core,
            metaData = Sign.Model.AppMetaData(
                name = metadata.name,
                description = metadata.description,
                url = metadata.url,
                icons = metadata.icons
            )
        )
        signClient = sign
        sign.initialize()

        // Subscribe to SDK events
        subscribeToSdkEvents(sign)

        isInitialized = true
    }

    /**
     * Subscribe to WalletConnectKotlin SDK events.
     */
    private fun subscribeToSdkEvents(sign: SignClient) {
        // Session settle (established)
        sessionSettleListenerId = sign.setSessionSettleListener { session ->
            val wcSession = session.toWCSession()
            scope.launch {
                _session.value = wcSession
                _accounts.value = wcSession.accounts
                _status.value = WCStatus.Connected
                _events.emit(WCEvent.Connected(wcSession))
            }
        }

        // Session delete (disconnect)
        sessionDeleteListenerId = sign.setSessionDeleteListener { topic ->
            scope.launch {
                _session.value = null
                _accounts.value = emptyList()
                _pairingUri.value = null
                _status.value = WCStatus.Disconnected
                _events.emit(WCEvent.Disconnected)
            }
        }

        // Session request (wallet-initiated RPC — rare for dApp)
        sessionRequestListenerId = sign.setSessionRequestListener { request ->
            // Handle wallet-initiated requests if needed
            scope.launch {
                _events.emit(WCEvent.SessionUpdate(session = _session.value))
            }
        }
    }

    /**
     * Configure the WC client.
     */
    fun configure(
        relayUrl: String,
        projectId: String,
        metadata: AppMetadata,
        chains: List<String> = listOf("eip155:1")
    ) {
        this.relayUrl = relayUrl
        this.projectId = projectId
        this.metadata = metadata
        this.requiredChains = chains
    }

    /**
     * Create a new pairing and generate a WC v2 URI.
     * Returns a WalletConnect URI string for QR display or deep linking.
     */
    suspend fun createPairing(): String = withContext(Dispatchers.IO) {
        val sign = signClient ?: throw WCError.NotConfigured

        // Build required namespaces for the session proposal
        val namespaces = buildRequiredNamespaces()

        // Create pairing via the SDK
        val pairingUri = sign.pair(
            requiredNamespaces = namespaces,
            optionalNamespaces = emptyMap(),
            sessionProperties = emptyMap(),
            topic = null,
            relay = Sign.Model.Relay("waku")
        )

        _pairingUri.value = pairingUri
        _status.value = WCStatus.Pairing

        pairingUri
    }

    /**
     * Connect using an existing WC v2 URI (e.g., from QR scan).
     */
    suspend fun connect(uri: String): WCSession = withContext(Dispatchers.IO) {
        val sign = signClient ?: throw WCError.NotConfigured

        // Parse URI to validate
        val parsed = parseWcUri(uri)

        _status.value = WCStatus.Connecting

        // Approve the pairing (SDK handles the session establishment)
        sign.pair(uri = uri)

        // Wait for session to settle
        waitForSession()
    }

    /**
     * Disconnect the current session.
     */
    suspend fun disconnect() {
        val sign = signClient ?: return
        val sessionVal = _session.value ?: return

        try {
            sign.disconnectSession(sessionVal.topic)
        } catch (_: Exception) {
            // Session may already be gone
        }

        _session.value = null
        _accounts.value = emptyList()
        _pairingUri.value = null
        _status.value = WCStatus.Disconnected
        _events.emit(WCEvent.Disconnected)
    }

    /**
     * Send a JSON-RPC request to the connected wallet.
     */
    suspend fun request(method: String, params: JSONArray): JSONObject = withContext(Dispatchers.IO) {
        val sign = signClient ?: throw WCError.NotConnected
        val sessionVal = _session.value ?: throw WCError.NotConnected

        val id = nextRequestId++
        val deferred = CompletableDeferred<JSONObject>()
        pendingRequests[id] = deferred

        // Timeout
        scope.launch {
            delay(60_000)
            pendingRequests.remove(id)?.completeExceptionally(WCError.RequestTimeout(method))
        }

        // Use the SDK's request method
        sign.request(
            topic = sessionVal.topic,
            method = method,
            params = params.toString(),
            chainId = sessionVal.accounts.firstOrNull()?.substringBefore(":")?.replace("eip155:", "eip155:") ?: "eip155:1"
        )

        deferred.await()
    }

    /**
     * Send eth_sendTransaction.
     */
    suspend fun sendTransaction(tx: WCTransactionRequest): String {
        val params = JSONArray().apply { put(tx.toJson()) }
        val result = request(WCMethods.ETH_SEND_TRANSACTION, params)
        return result.getString("result")
    }

    /**
     * Send personal_sign.
     */
    suspend fun personalSign(message: String, address: String): String {
        val hexMessage = if (message.startsWith("0x")) message else message.toHex()
        val params = JSONArray().apply {
            put(hexMessage)
            put(address)
        }
        val result = request(WCMethods.PERSONAL_SIGN, params)
        return result.getString("result")
    }

    /**
     * Fetch the connected account's native balance via eth_getBalance.
     */
    suspend fun fetchBalance(): String = withContext(Dispatchers.IO) {
        val accountsVal = _accounts.value
        if (accountsVal.isEmpty()) throw WCError.NotConnected

        // Extract address from CAIP-10 account string
        val caip10 = accountsVal.first()
        val address = caip10.substringAfterLast(":")

        // eth_getBalance via WC session request
        val params = JSONArray().apply {
            put(address)
            put("latest")
        }

        val result = request(WCMethods.ETH_GET_BALANCE, params)
        val balanceHex = result.getString("result")

        // Convert hex wei to decimal ETH
        val wei = balanceHex.hexToBigInteger()
        val eth = wei.toBigDecimal().divide(BigDecimal("1000000000000000000"), 4, java.math.RoundingMode.HALF_UP)
        eth.toPlainString()
    }

    /**
     * Switch chain.
     */
    suspend fun switchChain(chainId: Int) {
        val hexChainId = "0x" + Integer.toHexString(chainId)
        val params = JSONArray().apply {
            put(JSONObject().apply { put("chainId", hexChainId) })
        }
        request(WCMethods.WALLET_SWITCH_ETHEREUM_CHAIN, params)
    }

    // --- Private methods ---

    /**
     * Build the required namespaces map for session proposal.
     */
    private fun buildRequiredNamespaces(): Map<String, Sign.Model.Namespace> {
        return mapOf(
            "eip155" to Sign.Model.Namespace(
                chains = requiredChains,
                methods = requiredMethods,
                events = requiredEvents
            )
        )
    }

    /**
     * Wait for session establishment via the SDK event flow.
     */
    private suspend fun waitForSession(): WCSession = suspendCancellableCoroutine { continuation ->
        val job = scope.launch {
            events.collect { event ->
                when (event) {
                    is WCEvent.Connected -> {
                        continuation.resume(event.session) { /* cancelled */ }
                    }
                    is WCEvent.Error -> {
                        continuation.resumeWith(Result.failure(event.error))
                    }
                    else -> {}
                }
            }
        }

        // 5-minute timeout
        scope.launch {
            delay(300_000) // 5 minutes
            if (!continuation.isCompleted) {
                continuation.resumeWith(Result.failure(WCError.SessionTimeout))
            }
        }

        continuation.invokeOnCancellation { job.cancel() }
    }

    /**
     * Convert SDK Session model to our WCSession.
     */
    private fun Sign.Model.Session.toWCSession(): WCSession {
        val accountsList = namespaces.values.flatMap { namespace ->
            namespace.chains.flatMap { chain ->
                namespace.accounts.filter { it.startsWith(chain) }
            }
        }

        return WCSession(
            topic = topic,
            peerMetadata = peerMetadata?.let { mapOf("name" to it.name, "url" to it.url, "description" to it.description) } ?: emptyMap(),
            accounts = accountsList,
            namespaces = namespaces.mapValues { it.value },
            relayProtocol = relay.protocol
        )
    }
}

// ============================================================
// WC Utils — Crypto and URI helpers (for SDK-agnostic operations)
// ============================================================

object WCUtils {

    /**
     * Parse a WalletConnect v2 URI.
     */
    fun parseWcUri(uri: String): ParsedWcUri {
        val withoutPrefix = if (uri.startsWith("wc:")) uri.substring(3) else uri
        val parts = withoutPrefix.split("?", limit = 2)
        val topicVersion = parts[0].split("@")
        val topic = topicVersion[0]
        val version = topicVersion.getOrNull(1)?.toIntOrNull()
            ?: throw IllegalArgumentException("Invalid WC URI version")
        if (version != 2) throw IllegalArgumentException("Unsupported WC version: $version")

        val params = if (parts.size > 1) {
            parts[1].split("&").associate {
                val (k, v) = it.split("=", limit = 2)
                k to java.net.URLDecoder.decode(v, "UTF-8")
            }
        } else emptyMap()

        return ParsedWcUri(
            topic = topic,
            relayProtocol = params["relay-protocol"] ?: "waku",
            relayUrl = params["relay-url"] ?: "",
            symKey = params["symKey"] ?: throw IllegalArgumentException("Missing symKey")
        )
    }
}

// ============================================================
// Extensions
// ============================================================

private fun String.toHex(): String =
    "0x" + this.toByteArray(Charsets.UTF_8).joinToString("") { "%02x".format(it) }

private fun String.hexToBigInteger(): BigInteger {
    val cleaned = if (startsWith("0x")) substring(2) else this
    return BigInteger(cleaned, 16)
}

File: packages/ios-swift/Sources/OnChainUX/WalletConnect/WCClient.swift
/**
 * WCClient — WalletConnect v2 client for iOS using WalletConnectSwiftV2 SDK.
 *
 * Provides a native Swift implementation of the WalletConnect v2 protocol:
 * - Pairing URI generation for QR display
 * - Session proposal and establishment via WalletConnectSwiftV2
 * - X25519 key exchange (handled by SDK)
 * - Relay connection (Waku, handled by SDK)
 * - JSON-RPC dispatch (eth_sendTransaction, personal_sign, etc.)
 * - SIWE signing flow
 * - Balance fetching via on-chain RPC
 *
 * Uses the official WalletConnectSwiftV2 SDK via SPM.
 */

import Foundation
import Combine
import WalletConnect

// MARK: - WCClient

/// WalletConnect v2 client wrapping the official WalletConnectSwiftV2 SDK.
public final class WCClient: ObservableObject {
    
    /// Shared singleton instance.
    public static let shared = WCClient()
    
    // MARK: - Published State
    
    /// Current connection status.
    @Published public private(set) var status: WCStatus = .disconnected
    
    /// Active pairing URI for QR display.
    @Published public private(set) var pairingUri: String?
    
    /// Active session topic.
    @Published public private(set) var sessionTopic: String?
    
    /// Connected account addresses (CAIP-10).
    @Published public private(set) var accounts: [String] = []
    
    /// Connected chain ID.
    @Published public private(set) var chainId: Int = 1
    
    /// Peer wallet metadata.
    @Published public private(set) var peerMetadata: [String: String] = [:]
    
    // MARK: - Configuration
    
    /// WalletConnect projectId (from WalletConnect Cloud).
    public var projectId: String
    
    /// App metadata for session proposals.
    public var metadata: AppMetadata
    
    /// Required chains (CAIP-2 format, e.g. "eip155:1").
    public var requiredChains: [String]
    
    /// Required methods.
    public var requiredMethods: [String]
    
    /// Required events.
    public var requiredEvents: [String]
    
    // MARK: - Internal WalletConnectSwiftV2 References
    
    /// The NetworkingInteractor from the SDK.
    internal var networking: NetworkingInteractor?
    
    /// The Sign client for session management.
    internal var signClient: Sign?
    
    /// The Pairing client.
    internal var pairingClient: Pair?
    
    /// Combine cancellables.
    internal var cancellables = Set<AnyCancellable>()
    
    /// Event handlers.
    private var eventHandlers: [UUID: (WCEvent) -> Void] = [:]
    
    /// Pending request callbacks.
    private var pendingRequests: [UInt64: CheckedContinuation<Any, Error>] = [:]
    
    /// Next JSON-RPC request ID.
    private var nextRequestId: UInt64 = 1
    
    private init(
        projectId: String = "",
        metadata: AppMetadata = .default,
        requiredChains: [String] = ["eip155:1"],
        requiredMethods: [String] = WCMethods.standardEvmMethods,
        requiredEvents: [String] = WCEvents.standardEvmEvents
    ) {
        self.projectId = projectId
        self.metadata = metadata
        self.requiredChains = requiredChains
        self.requiredMethods = requiredMethods
        self.requiredEvents = requiredEvents
    }
    
    // MARK: - Configuration
    
    /// Configure the WC client with app settings.
    public func configure(
        projectId: String,
        metadata: AppMetadata = .default,
        chains: [String] = ["eip155:1"]
    ) {
        self.projectId = projectId
        self.metadata = metadata
        self.requiredChains = chains
    }
    
    // MARK: - SDK Initialization
    
    /// Initialize the WalletConnectSwiftV2 SDK internals.
    /// Call this once at app launch before any WC operations.
    public func initializeSDK(
        metadata: AppMetadata,
        projectId: String,
        isController: Bool = true
    ) {
        // Configure the SDK's NetworkingInteractor
        let networking = NetworkingInteractor.configure(
            projectId: projectId,
            socketFactory: DefaultSocketFactory()
        )
        self.networking = networking
        
        // Configure Pairing client
        self.pairingClient = Pair.configure(networking: networking)
        
        // Configure Sign client
        self.signClient = Sign.configure(
            networking: networking,
            keyValueStorage: KeyValueStorageUserDefaults.standard
        )
        
        // Listen for session proposals (wallet-side; dApp side may use different flow)
        signClient?.sessionProposalPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] proposal in
                self?.handleSessionProposal(proposal)
            }
            .store(in: &cancellables)
        
        // Listen for session requests
        signClient?.sessionRequestPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] request in
                self?.handleSessionRequest(request)
            }
            .store(in: &cancellables)
        
        // Listen for session state changes
        signClient?.sessionDeletePublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in
                self?.sessionTopic = nil
                self?.accounts = []
                self?.status = .disconnected
                self?.emit(.disconnected)
            }
            .store(in: &cancellables)
        
        self.metadata = metadata
        self.projectId = projectId
    }
    
    // MARK: - Pairing
    
    /// Create a new pairing and generate a WC v2 URI for QR display.
    public func createPairing() async throws -> String {
        guard let pairing = pairingClient else {
            throw WCError.notConfigured
        }
        
        let uri = try await pairing.create()
        self.pairingUri = uri.absoluteString
        status = .pairing
        
        return uri.absoluteString
    }
    
    /// Connect using an existing WC v2 URI (e.g., from QR scan).
    public func connect(uri: String) async throws -> WCSession {
        guard let pairing = pairingClient else {
            throw WCError.notConfigured
        }
        
        status = .connecting
        
        // Pair with the URI
        let pairingInfo = try await pairing.pair(uri: uri)
        
        // Wait for session to be established
        return try await waitForSession(topic: pairingInfo.topic)
    }
    
    /// Disconnect the current session.
    public func disconnect() async {
        guard let sign = signClient, let topic = sessionTopic else { return }
        
        try? await sign.disconnect(
            topic: topic,
            reason: .userDisconnected
        )
        
        sessionTopic = nil
        accounts = []
        pairingUri = nil
        peerMetadata = [:]
        status = .disconnected
        emit(.disconnected)
    }
    
    // MARK: - Session Management
    
    /// Handle an incoming session proposal (wallet side).
    private func handleSessionProposal(_ proposal: Session.Proposal) {
        Task {
            do {
                guard let sign = signClient else { return }
                
                let namespaces: [String: SessionNamespace] = [
                    "eip155": SessionNamespace(
                        accounts: requiredChains.map { chain in
                            "\(chain):0x0000000000000000000000000000000000000000"
                        },
                        methods: requiredMethods,
                        events: requiredEvents
                    )
                ]
                
                let session = try await sign.approve(
                    proposalId: proposal.id,
                    namespaces: namespaces
                )
                
                sessionTopic = session.topic
                status = .connected
                emit(.connected(session: WCSession(topic: session.topic)))
            } catch {
                status = .error(error.localizedDescription)
                emit(.error(WCError.relayError(error.localizedDescription)))
            }
        }
    }
    
    /// Handle an incoming session request.
    private func handleSessionRequest(_ request: Request) {
        print("[WCClient] Received session request: \(request.method)")
    }
    
    /// Wait for session establishment after pairing.
    private func waitForSession(topic: String) async throws -> WCSession {
        try await withCheckedThrowingContinuation { continuation in
            guard let sign = signClient else {
                continuation.resume(throwing: WCError.notConfigured)
                return
            }
            
            sign.sessionSettlePublisher
                .receive(on: DispatchQueue.main)
                .first { $0.topic == topic }
                .sink { session in
                    self.sessionTopic = session.topic
                    
                    var sessionAccounts: [String] = []
                    for namespace in session.namespaces.values {
                        sessionAccounts.append(
                            contentsOf: namespace.accounts.map { $0.absoluteString }
                        )
                    }
                    
                    self.accounts = sessionAccounts
                    if let first = sessionAccounts.first {
                        let parts = first.split(separator: ":")
                        if parts.count >= 2, let cid = Int(parts[1]) {
                            self.chainId = cid
                        }
                    }
                    
                    self.status = .connected
                    self.emit(.connected(
                        session: WCSession(topic: topic, accounts: sessionAccounts)
                    ))
                    
                    continuation.resume(returning: WCSession(
                        topic: topic,
                        accounts: sessionAccounts
                    ))
                }
                .store(in: &self.cancellables)
            
            // 5-minute timeout
            Task {
                try? await Task.sleep(nanoseconds: 300_000_000_000)
                if self.sessionTopic != topic {
                    continuation.resume(throwing: WCError.sessionTimeout)
                }
            }
        }
    }
    
    // MARK: - JSON-RPC Requests
    
    /// Send a JSON-RPC request to the connected wallet.
    public func request<T: Decodable>(method: String, params: Any) async throws -> T {
        guard let sign = signClient, let topic = sessionTopic else {
            throw WCError.notConnected
        }
        
        let request = Request(
            topic: topic,
            method: method,
            params: params,
            chainId: Blockchain("eip155:\(chainId)")!
        )
        
        return try await sign.request(params: request)
    }
    
    /// Convenience: send eth_sendTransaction.
    public func sendTransaction(_ tx: WCTransactionRequest) async throws -> String {
        let params: [[String: String]] = [[
            "from": tx.from,
            "to": tx.to,
            "value": tx.value ?? "0x0",
            "data": tx.data ?? "0x",
            "gas": tx.gas ?? "0x5208",
        ].compactMapValues { $0 }]
        return try await request(method: WCMethods.ethSendTransaction, params: params)
    }
    
    /// Convenience: send personal_sign.
    public func personalSign(message: String, address: String) async throws -> String {
        let hexMessage = message.hasPrefix("0x") ? message : message.utf8Hex
        return try await request(
            method: WCMethods.personalSign,
            params: [hexMessage, address]
        )
    }
    
    /// Convenience: send eth_signTypedData_v4 (EIP-712).
    public func signTypedData(address: String, typedData: String) async throws -> String {
        try await request(
            method: WCMethods.ethSignTypedDataV4,
            params: [address, typedData]
        )
    }
    
    /// Convenience: switch chain.
    public func switchChain(chainId: Int) async throws {
        let hexChainId = "0x" + String(chainId, radix: 16)
        try await request(
            method: WCMethods.walletSwitchEthereumChain,
            params: [["chainId": hexChainId]]
        )
    }
    
    // MARK: - Balance Fetching
    
    /// Fetch the native balance for the connected account via eth_getBalance.
    public func fetchBalance() async throws -> String {
        guard !accounts.isEmpty else {
            throw WCError.notConnected
        }
        
        let address = accounts[0].split(separator: ":").last.map(String.init) ?? accounts[0]
        
        let balanceHex: String = try await request(
            method: "eth_getBalance",
            params: [address, "latest"]
        )
        
        // Convert hex wei to decimal ETH
        let cleaned = balanceHex.replacingOccurrences(of: "0x", with: "")
        if let wei = BigInt(cleaned, radix: 16) {
            let ethWei = Decimal(wei.magnitude)
            let eth = ethWei / Decimal(10).power(18)
            return String(format: "%.4f", (eth as NSDecimalNumber).doubleValue)
        }
        
        return "0.0000"
    }
    
    // MARK: - Events
    
    /// Subscribe to WC client events.
    public func onEvent(_ handler: @escaping (WCEvent) -> Void) -> UUID {
        let id = UUID()
        eventHandlers[id] = handler
        return id
    }
    
    /// Unsubscribe from events.
    public func unsubscribe(_ id: UUID) {
        eventHandlers.removeValue(forKey: id)
    }
    
    /// Emit an event to all handlers.
    private func emit(_ event: WCEvent) {
        for handler in eventHandlers.values {
            handler(event)
        }
    }
}

// MARK: - Types

/// WC client connection status.
public enum WCStatus: Equatable {
    case disconnected
    case connecting
    case pairing
    case connected
    case error(String)
    
    public static func == (lhs: WCStatus, rhs: WCStatus) -> Bool {
        switch (lhs, rhs) {
        case (.disconnected, .disconnected): return true
        case (.connecting, .connecting): return true
        case (.pairing, .pairing): return true
        case (.connected, .connected): return true
        case let (.error(l), .error(r)): return l == r
        default: return false
        }
    }
}

/// WC v2 session data.
public struct WCSession: Sendable {
    public let topic: String
    public let peerMetadata: [String: Any]
    public let accounts: [String]
    public let namespaces: [String: Any]
    public let relay: [String: Any]
    
    public init(
        topic: String,
        peerMetadata: [String: Any] = [:],
        accounts: [String] = [],
        namespaces: [String: Any] = [:],
        relay: [String: Any] = [:]
    ) {
        self.topic = topic
        self.peerMetadata = peerMetadata
        self.accounts = accounts
        self.namespaces = namespaces
        self.relay = relay
    }
}

/// Application metadata for WC session proposals.
public struct WCAppMetadata: Sendable {
    public let name: String
    public let description: String
    public let url: String
    public let icons: [String]
    
    public init(name: String, description: String, url: String, icons: [String]) {
        self.name = name
        self.description = description
        self.url = url
        self.icons = icons
    }
    
    public static let `default` = WCAppMetadata(
        name: "Cinacoin dApp",
        description: "Connected via Cinacoin",
        url: "https://cinacoin.io",
        icons: ["https://cinacoin.io/icon.png"]
    )
}

/// Transaction request for eth_sendTransaction.
public struct WCTransactionRequest: Sendable {
    public let from: String
    public let to: String
    public let value: String?
    public let data: String?
    public let gas: String?
    public let gasPrice: String?
    public let maxFeePerGas: String?
    public let maxPriorityFeePerGas: String?
    public let nonce: String?
    public let chainId: Int?
    
    public init(
        from: String,
        to: String,
        value: String? = nil,
        data: String? = nil,
        gas: String? = nil,
        gasPrice: String? = nil,
        maxFeePerGas: String? = nil,
        maxPriorityFeePerGas: String? = nil,
        nonce: String? = nil,
        chainId: Int? = nil
    ) {
        self.from = from
        self.to = to
        self.value = value
        self.data = data
        self.gas = gas
        self.gasPrice = gasPrice
        self.maxFeePerGas = maxFeePerGas
        self.maxPriorityFeePerGas = maxPriorityFeePerGas
        self.nonce = nonce
        self.chainId = chainId
    }
}

/// WC client events.
public enum WCEvent {
    case pairingCreated(uri: String)
    case sessionProposal([String: Any])
    case connected(session: WCSession)
    case sessionUpdate(session: WCSession?)
    case disconnected
    case error(WCError)
}

/// WC-specific errors.
public enum WCError: Error, LocalizedError {
    case notConnected
    case notConfigured
    case invalidUrl(String)
    case invalidUri(String)
    case sessionTimeout
    case requestTimeout(String)
    case rpcError(Int, String)
    case invalidProposalResponse
    case relayError(String)
    case cryptoError(String)
    
    public var errorDescription: String? {
        switch self {
        case .notConnected: return "Not connected to a wallet"
        case .notConfigured: return "WCClient has not been configured"
        case let .invalidUrl(url): return "Invalid relay URL: \(url)"
        case let .invalidUri(uri): return "Invalid WalletConnect URI: \(uri)"
        case .sessionTimeout: return "Session establishment timed out"
        case let .requestTimeout(method): return "Request '\(method)' timed out"
        case let .rpcError(code, message): return "RPC error \(code): \(message)"
        case .invalidProposalResponse: return "Invalid session proposal response"
        case let .relayError(msg): return "Relay error: \(msg)"
        case let .cryptoError(msg): return "Crypto error: \(msg)"
        }
    }
}

// MARK: - Standard Methods & Events

/// Standard WC v2 methods for EVM chains.
public enum WCMethods {
    public static let ethSendTransaction = "eth_sendTransaction"
    public static let ethSignTransaction = "eth_signTransaction"
    public static let personalSign = "personal_sign"
    public static let ethSignTypedData = "eth_signTypedData"
    public static let ethSignTypedDataV4 = "eth_signTypedData_v4"
    public static let walletSwitchEthereumChain = "wallet_switchEthereumChain"
    public static let walletAddEthereumChain = "wallet_addEthereumChain"
    public static let ethAccounts = "eth_accounts"
    public static let ethChainId = "eth_chainId"
    
    public static let standardEvmMethods: [String] = [
        ethSendTransaction, ethSignTransaction, personalSign,
        ethSignTypedData, ethSignTypedDataV4,
        walletSwitchEthereumChain, walletAddEthereumChain,
        ethAccounts, ethChainId
    ]
}

/// Standard WC v2 events for EVM chains.
public enum WCEvents {
    public static let chainChanged = "chainChanged"
    public static let accountsChanged = "accountsChanged"
    public static let standardEvmEvents: [String] = [chainChanged, accountsChanged]
}

// MARK: - BigInt Helper

struct BigInt: Equatable {
    let magnitude: UInt64
    
    init?(_ string: String, radix: Int) {
        guard radix == 16 else { return nil }
        let cleaned = string.replacingOccurrences(of: "0x", with: "")
        guard let val = UInt64(cleaned, radix: 16) else { return nil }
        self.magnitude = val
    }
}

// MARK: - String Extensions

extension String {
    var urlEncoded: String {
        addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? self
    }
    
    var utf8Hex: String {
        "0x" + self.utf8.map { String(format: "%02x", $0) }.joined()
    }
}
